### PR TITLE
SOLR-16265 reduce memory usage of ContentWriter based requests in Http2SolrClient

### PR DESCRIPTION
**Pull request by stillalex**
https://issues.apache.org/jira/browse/SOLR-16265

# Description

Initially started with addressing the TODO to remove extra buffer from the code (https://github.com/apache/solr/pull/1789/files#diff-db379c74efc9ff28acfd7571b07e16dbcdaa510783f2bee835b4a288927b5b3dR694). Ended up with removing all deprecated code to refresh implementation and bring it up to latest guidance.
It looks a bit more involved because some of the writers now need to push content after being setup, opposed to before style that was upfront setup of content.

To evaluate some performance aspect I ran exiting indexing benchmark. Numbers were equivalent, but heap footprint on benchmark client side saw a drop. posted screenshots below. if there are other tests I could run please let me know.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by stillalex** [comment]
Attaching memory footprint of client side during indexing benchmark. While perf numbers are very similar, memory footprint definitely sees a drop.

* main branch

bench-http2client-main

* proposed patch

bench-http2client-patch

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 1]
```diff
@@ -755,11 +762,12 @@ private Request fillContentStream(
           content.addFilePart(
               name,
               contentStream.getName(),
-              new InputStreamContentProvider(contentStream.getStream()),
+              new InputStreamRequestContent(contentStream.getStream()),
               fields);
         }
       }
-      req.content(content);
+      content.close();
```
Comment (risdenk) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:769]: Do we have to worry about closing this if there is an exception? not sure if `MultiPartRequestContent` can do try w/ resources?

**Review by risdenk — APPROVED** [review]

**Comment by epugh** [comment]
I wish we had more confidence in our tests (benchmark and unit) that a change like this that passed the tests could be assumed to correct ;-).   I like anything that removes deprecated code!

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java (hunk 1 above)
Comment (stillalex) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:769]: good idea! seems it does so I will give it a go.

**Comment by stillalex** [comment]
> I wish we had more confidence in our tests (benchmark and unit) that a change like this that passed the tests could be assumed to correct ;-). I like anything that removes deprecated code!

oh yeah, I completely agree. how can we increase that confidence? I feel like there are quite a few tests already (running 'gradlew check' is already a lengthy affair)

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java (hunk 1 above)
Comment (stillalex) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java]: change is in!

**Comment by stillalex** [comment]
thank you @risdenk and @markrmiller for the reviews. I have added one more change which was a bit awkward. I needed the request url for the exception message inside the async listener, so I moved some methods around to build the url first. it's the least bad solution I could come up with. this doesn't change much of the PR, so I will go ahead and try to push this in once the builds complete.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 2]
```diff
@@ -438,81 +436,78 @@ public void send(OutStream outStream, SolrRequest<?> req, String collection) thr
...
-                        NamedList<Object> body =
-                            processErrorsAndResponse(
-                                solrRequest, parser, response, is, req.getURI().toString());
+      String url = getRequestPath(solrRequest, collection);
+      InputStreamResponseListener listener =
+          new InputStreamResponseListener() {
+            @Override
+            public void onHeaders(Response response) {
+              super.onHeaders(response);
+              InputStreamResponseListener listener = this;
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:451]: must this be defined a second time?

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 3]
```diff
@@ -659,62 +633,97 @@ private Request createRequest(SolrRequest<?> solrRequest, String collection)
...
-        Request req = httpClient.newRequest(url + wparams.toQueryString()).method(method);
-        Utils.BAOS baos = new Utils.BAOS();
-        contentWriter.write(baos);
-
-        // SOLR-16265: TODO reduce memory usage
-        return req.content(
-            // We're throwing this BAOS away, so no need to copy the byte[], just use the raw buf
-            new ByteBufferContentProvider(
-                contentWriter.getContentType(), ByteBuffer.wrap(baos.getbuf(), 0, baos.size())));
+        OutputStreamRequestContent content =
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:695]: Just want to mention that cases like this are perfect for "var".  For next time...

**Review by dsmiley — APPROVED** [review]
Love the perf analysis!

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java (hunk 3 above)
Comment (stillalex) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java]: agreed. I added a few of those below, this one got away. replaced it with latest commit

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java (hunk 2 above)
Comment (stillalex) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java]: does not seem to be needed. weird, removed it now

**Comment by stillalex** [comment]
will push changes.txt to trigger a new build (I needed to do that anyway)

**Pull request merged**